src/qibolab/instruments/zhinst/sweep.py

```diff
@@ -0,0 +1,139 @@
+"""Pre-execution processing of sweeps."""
+
+from collections.abc import Iterable
+from copy import copy
+
+import laboneq.simple as lo
+
+from qibolab.pulses import Pulse, PulseType
+from qibolab.qubits import Qubit
+from qibolab.sweeper import Parameter, Sweeper
+
+from .util import NANO_TO_SECONDS, measure_channel_name
+
+
+def classify_sweepers(
+    sweepers: Iterable[Sweeper],
+) -> tuple[list[Sweeper], list[Sweeper]]:
+    """Divide sweepers into two lists: 1. sweeps that can be done in the laboneq near-time sweep loop, 2. sweeps that
+    can be done in real-time (i.e. on hardware)"""
+    nt_sweepers, rt_sweepers = [], []
+    for sweeper in sweepers:
+        if (
+            sweeper.parameter is Parameter.amplitude
+            and sweeper.pulses[0].type is PulseType.READOUT
+        ):
+            nt_sweepers.append(sweeper)
+        elif sweeper.parameter is Parameter.bias:
+            nt_sweepers.append(sweeper)
+        else:
+            rt_sweepers.append(sweeper)
+    return nt_sweepers, rt_sweepers
+
+
+class ProcessedSweeps:
+    """Data type that centralizes and allows extracting information about given
+    sweeps."""
+
+    def __init__(self, sweepers: Iterable[Sweeper], qubits: dict[str, Qubit]):
+        pulse_sweeps = []
+        channel_sweeps = []
+        parallel_sweeps = []
+        for sweeper in sweepers:
+            for pulse in sweeper.pulses or []:
+                if sweeper.parameter in (Parameter.duration, Parameter.start):
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values * NANO_TO_SECONDS
+                    )
+                    pulse_sweeps.append((pulse, sweeper.parameter, sweep_param))
+                elif sweeper.parameter is Parameter.frequency:
+                    ptype, qubit = pulse.type, qubits[pulse.qubit]
+                    if ptype is PulseType.READOUT:
+                        ch = measure_channel_name(qubit)
+                        intermediate_frequency = (
+                            qubit.readout_frequency
+                            - qubit.readout.local_oscillator.frequency
+                        )
+                    elif ptype is PulseType.DRIVE:
+                        ch = qubit.drive.name
+                        intermediate_frequency = (
+                            qubit.drive_frequency
+                            - qubit.drive.local_oscillator.frequency
+                        )
+                    else:
+                        raise ValueError(
+                            f"Cannot sweep frequency of pulse of type {ptype}, because it does not have associated frequency"
+                        )
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values + intermediate_frequency
```

Comment:
This is also a bit weird for me: why do you have to sum the IF?

At most, I expected you had to subtract the LO freq, to move it out of the RF band.

I'm for sure msising something...

Reply:
I don't know what is the full history behind this, but my understanding is that this is one of those places where the driver assumes things about the sweep that it shouldn't. Namely, it silently assumes that the given sweep is of type `SweeperType.OFFSET`, hence the sweeper values have to be added to the center frequency. The center frequency being the intermediate frequency in this case.

Reply:
Ok, I see. This is potentially harmful, and we could benchmark what is happening as Qibocal users (or mock-Qibocal, implementing pseudo-routines ourselves if Qibocal is not discriminating enough).

Reply:
> I don't know what is the full history behind this, but my understanding is that this is one of those places where the driver assumes things about the sweep that it shouldn't. Namely, it silently assumes that the given sweep is of type `SweeperType.OFFSET`, hence the sweeper values have to be added to the center frequency. The center frequency being the intermediate frequency in this case.

Yes, you are right about the assumption.